### PR TITLE
Fix: data leakage was incorrect with window functions BREAKING CHANGE

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/core/audit/definition.py [hunk 1]
```diff
@@ -184,16 +184,20 @@ def render_query(
...
-        query_renderer.filter_time_column(this_model_subquery, start or c.EPOCH, end or c.EPOCH)
-
         return query_renderer.render(
             start=start,
             end=end,
             latest=latest,
             snapshots=snapshots,
             is_dev=is_dev,
-            this_model=this_model_subquery.subquery(),
+            this_model=exp.select("*")
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:193]: IMHO this is a lot for an inline argument. Can we put this into a variable?

File: sqlmesh/core/renderer.py [hunk 2]
```diff
@@ -235,13 +234,19 @@ def _expand(node: exp.Expression) -> exp.Expression:
...
         # events that have data outside the time window of interest.
-        if add_incremental_filter:
-            # expansion copies the query for us. if it doesn't occur, make sure to copy.
-            if not expand:
-                query = query.copy()
-            for node, _, _ in query.walk(prune=lambda n, *_: isinstance(n, exp.Select)):
-                if isinstance(node, exp.Select):
-                    self.filter_time_column(node, *dates[0:2])
+        if add_incremental_filter and self._time_column:
+            "Filters a query on the time column to ensure no data leakage when running in incremental mode."
```
Comment (GeorgeSittas) [inline sqlmesh/core/renderer.py:238]:
```suggestion
            """Filters a query on the time column to ensure no data leakage when running in incremental mode."""
```

Comment (izeigerman) [inline sqlmesh/core/renderer.py:238]: Is this a string literal? Should this be a comment?

File: sqlmesh/core/renderer.py [hunk 3]
```diff
@@ -235,13 +234,19 @@ def _expand(node: exp.Expression) -> exp.Expression:
...
+            # expand does a copy already
+            query = t.cast(exp.Subqueryable, query if expand else query.copy())
+            with_ = query.args.pop("with", None)
+            query = (
+                exp.select(*(["*"] if query.is_star else query.named_selects))
```
Comment (izeigerman) [inline sqlmesh/core/renderer.py:243]: Why can't it always be `*`?

File: sqlmesh/core/renderer.py [hunk 4]
```diff
@@ -251,6 +256,17 @@ def _expand(node: exp.Expression) -> exp.Expression:
 
         return t.cast(exp.Subqueryable, query)
 
+    def time_column_filter(self, start: TimeLike, end: TimeLike) -> exp.Between:
+        """Returns a between statement with the properly formatted time column."""
+        if not self._time_column:
+            raise RuntimeError(
```
Comment (izeigerman) [inline sqlmesh/core/renderer.py:262]: `SQLMeshError` maybe

File: sqlmesh/core/renderer.py [hunk 5]
```diff
@@ -251,6 +256,17 @@ def _expand(node: exp.Expression) -> exp.Expression:
...
+                f"Cannot produce time column filter because model does not have a time column."
+            )
+
+        return exp.column(self._time_column.column)[
+            self._time_converter(start) : self._time_converter(end)  # type: ignore
```
Comment (izeigerman) [inline sqlmesh/core/renderer.py:267]: Please don't tell me this is the "between" DSL

Comment (izeigerman) [reply sqlmesh/core/renderer.py:267]: Please note that in Python the end index is exclusive so it doesn't map well to BETWEEN. Other issues with it aside.

Comment (izeigerman) [reply sqlmesh/core/renderer.py:267]: Can't this just be:
```
exp.column(self._time_column.column).between(...)
```

File: sqlmesh/core/renderer.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/core/renderer.py:238]: lol woops

File: sqlmesh/core/renderer.py (hunk 3 above)
Comment (tobymao) [reply sqlmesh/core/renderer.py:243]: my first attempt did this pre column_types and so i needed this expanded, but i've removed that so i can move back to star

Comment (izeigerman) [reply sqlmesh/core/renderer.py:243]: I think it'd just remove unnecessary details from the code and will make it easier to read

File: sqlmesh/core/renderer.py (hunk 5 above)
Comment (tobymao) [reply sqlmesh/core/renderer.py:267]: will fix this once george's stuff is in

File: sqlmesh/core/model/kind.py [hunk 6]
```diff
@@ -80,6 +80,12 @@ class TimeColumn(PydanticModel):
     column: str
     format: t.Optional[str] = None
 
+    @validator("column", pre=True)
+    def _column_validator(cls, v: str) -> str:
+        if not v:
+            raise ConfigError(f"Time Column cannot be empty.")
```
Comment (vchan) [inline sqlmesh/core/model/kind.py:86]:
```suggestion
            raise ConfigError("Time Column cannot be empty.")
```

File: sqlmesh/core/renderer.py [hunk 7]
```diff
@@ -251,6 +260,17 @@ def _expand(node: exp.Expression) -> exp.Expression:
 
         return t.cast(exp.Subqueryable, query)
 
+    def time_column_filter(self, start: TimeLike, end: TimeLike) -> exp.Between:
+        """Returns a between statement with the properly formatted time column."""
+        if not self._time_column:
+            raise SQLMeshError(
+                f"Cannot produce time column filter because model does not have a time column."
```
Comment (vchan) [inline sqlmesh/core/renderer.py:267]:
```suggestion
                "Cannot produce time column filter because model does not have a time column."
```

**Pull request merged**